Add HTML conversion recipe (#268)
apostrophecms/a3-docs: docs/cookbook/html-conversion.md

# Converting a premade HTML template for use in an Apostrophe Project

Creating a web site from scratch can be a daunting process. Not only do you have to create the HTML markup, but also the styling to make it all look good. One way to get started quickly is to use one of the numerous premade templates available on the internet. In this tutorial, we will walk through the steps to convert the ["Start Bootstrap Clean Blog"](https://startbootstrap.com/theme/clean-blog) template into an Apostrophe template. While we are starting with a specific template, these steps can be generalized to any template you might download.

## Overview

Let's outline the steps that we need to perform.

1. Create a new boilerplate project
2. Add the styling and scripts from the template to our project

> 3. Identify sections found on each page that can be converted into fragments

Author: Good idea.

   - Navigation
   - Header
   - Footer
4. Modify the project core layout
5. Create the default page

> 6. Add the blog pages

Author: Good suggestions - I was struggling with how to differentiate the apostrophe layout versus the template.

   - create the show.html page

## Template Introduction

This template is free to [download](https://startbootstrap.com/theme/clean-blog) and contains 4 simple pages. The "Home" page contains a listing of all the blog articles on the site.
![The template home page](../.vuepress/public/images/home-page.png)

This has an accompanying page to show each individual article.

![The template 'show.html' page](../.vuepress/public/images/show-page.png)

This structure matches up nicely with the structure of the apostrophe `piece-page-type` with an `index.html` template to list all of the pieces and a `show.html` template to show each individual piece.

The final two pages are an "About" page and a "Contact Me" page, which have identical structure, just content differences.

![The 'Contact-me" page.'](../.vuepress/public/images/contact-me.png)

All of these pages have very similar headers containing fixed navigation, a large header image, and some text over the image.

They also all display an identical footer containing social links and some copyright text.

The styling of the template is a combination of the popular Bootstrap frontend styling framework and custom CSS.

> Let's get started converting this template to an Apostrophe project. The first step is to [download](https://startbootstrap.com/theme/clean-blog) the template and unarchive it to a directory where we can copy files into our Apostrophe project.

Author: Changed. You are right, it was too redundant. I did a little rewording and deletion.

## Creating a new project

> If you already have the apostrophe CLI app installed, creating a new project is simple. If not, follow the instructions [here](https://v3.docs.apostrophecms.org/guide/setting-up.html#the-apostrophe-cli-tool). Next, create a new project from the command line. Make sure you are in the directory where you want to create your new project folder and run the following command:

Author: Changed.

<AposCodeBlock>

```sh
apos create template-app
```

</AposCodeBlock>

The CLI app will create the new project and an admin user. At the end of the installation, it will ask for an administrator password - make sure to remember this password for login.

## Adding Bootstrap and project styling

### Adding the styling

> This particular template comes with both "dist" and "src" folders. Within the dist folder, is a "css" folder that contains all of the compiled styling for the site. We could use this as the source for the styling of our project, but this wouldn't be as easy to modify with additional or custom styling variables.

Author: Yeah, thanks.

> Alternatively, the src folder contains an "scss" folder with all of the styling sheets and imports. Since this template utilizes Bootstrap, which has a npm package, we are going to install and then include the main styling from the 'node-modules' folder. Open a terminal at the root of your project and install Bootstrap using:

Author: Changed.

```sh
npm install bootstrap
```

::: note
This template uses Bootstrap 5, which is the latest version as of this writing. If you need another version for your template, make sure to specify it during the install.
:::

> The next thing we will do is copy the contents of the scss folder into our project. There are multiple places that we could add these files, but the best place is in a project-level `modules/@postrophecms/asset` folder. If the folder doesn't already exist in your project, create it. Next, within that folder create a `ui/src` folder and copy the entirety of the `dist/scss` folder.

Author: Changed.

> For the HTML template, the `styles.scss` file is the entry point for loading all of the individual scss sheets. For our Apostrophe project, we are going to move this sheet up one level from the `/scss` folder into the `ui/src` folder and rename it `index.scss`. Next, we need to edit this file to point to all of the sub-sheets. Looking at the file path for each `@import` statement, each sub-sheet or folder of sub-sheets is expected to be found in the same folder as the entry sheet. After copying it into our project, this is no longer true. Instead, all of the sheets are located within the `scss` folder of the same directory. Modify all of the `@import` statements (except for the Bootstrap import) to point to the correct location by prefixing the path with the folder name:

Author: Changed.

<AposCodeBlock>

```scss
// Import variables
@import './scss/variables';

// import bootstrap
// This path is pointing to the Bootstrap package in the node_module folder

> @import 'bootstrap/scss/bootstrap.scss';

Author: Huh, not sure what to do because at that same link they say that using the tilde is deprecated.

> @import 'bootstrap/scss/bootstrap.scss';

Reviewer: Doh... Missed that. You're right. If it works without don't need it.

// Global CSS
@import './scss/global';

// Components
@import './scss/components/buttons';
@import './scss/components/forms';
@import './scss/components/navbar';

// Sections
@import './scss/sections/footer';
@import './scss/sections/masthead';
@import './scss/sections/post';
```

<template v-slot:caption>
modules/@apostrophecms/asset/ui/src/index.scss
</template>
</AposCodeBlock>

> The main Bootstrap components are loaded in from the `node_modules` where they were installed. An alternative to directly loading from the `node-modules`, which will load in every Bootstrap component, would be to create another folder and import only those components needed for the project.

Author: D'oh! Yes, yes you could.

Create another file named `index.js` within the `modules/@apostrophecms/asset/ui/src` folder. Within this file, we can import the main Javascript bundle and add the custom script from the template `src/js/scripts.js` file.

<AposCodeBlock>

```js
// import the bootstrap.js and popper.js files from the node_modules
import bootstrap from 'bootstrap';

export default () => {
  //code from the template
  window.addEventListener('DOMContentLoaded', () => {
    let scrollPos = 0;
    const mainNav = document.getElementById('mainNav');
    const headerHeight = mainNav.clientHeight;
    window.addEventListener('scroll', function () {
      const currentTop = document.body.getBoundingClientRect().top * -1;
      if (currentTop < scrollPos) {
        // Scrolling Up
        if (currentTop > 0 && mainNav.classList.contains('is-fixed')) {
          mainNav.classList.add('is-visible');
        } else {

>           console.log(123);

Author: Hah! That was in the original downloaded template - I just did a cut and paste. Deleted.

          mainNav.classList.remove('is-visible', 'is-fixed');
        }
      } else {
        // Scrolling Down
        mainNav.classList.remove(['is-visible']);
        if (
          currentTop > headerHeight &&
          !mainNav.classList.contains('is-fixed')
        ) {
          mainNav.classList.add('is-fixed');
        }
      }
      scrollPos = currentTop;
    });
  });
};
```

<template v-slot:caption>
modules/@apostrophecms/asset/ui/src/index.js
</template>

</AposCodeBlock>

## Identifying common areas

> Each of the four pages included in this template has some common areas that can be converted into fragments. While all of the pages have both navigation and main header areas that occupy the same general area, we are going to split these into two fragments. That is because the navigation can get all of its settings from global, whereas the rest of the header area is going to get settings on a per-page basis. We will add each of the three fragments in the `views` folder at the project level.

Author: Yup, yup.

### Adding the navigation

Inside the `views` folder create another folder named `fragments` and a file named `navigation.html`. To turn this page into a fragment, add opening and closing fragment block tags - `{% fragment navigationArea() %}` and `{% endfragment %}`.

> Open one of the template pages and copy the navigation section. Paste this between the two fragment tags. To add the website brand to the navigation, we will replace the href and logo in the first link with data from schema fields in the global settings. Next, within the unordered list, delete the last three `<li>` items. To populate the list with each of the pages we will use a `for` loop to add each page selected within the global settings.

Author: I back ticked the `href` and added 'apostrophe' before the first mention of global settings.

      homePage: {
        label: 'Page to link',
        type: 'relationship',
        withType: '@apostrophecms/page',
        max: 1,
        required: true,
        builders: {
          project: {
            title: 1,
            _url: 1
          }
        }
      },
      brand: {
        type: 'string',
        label: 'Brand name'
      },
      pages: {
        type: 'array',
        titleField: 'Pages',
        fields: {
          add: {
            label: {
              type: 'string',
              label: 'Page label'
            },
            _page: {
              label: 'Page to link',
              type: 'relationship',
              withType: '@apostrophecms/page',
              max: 1,
              required: true,
              builders: {
                project: {
                  title: 1,
                  _url: 1
                }
              }
            }
          }
        }
      }
    },
    group: {
      navigation: {
        label: 'Navigation links',
        fields: ['homePage', 'brand', 'pages']
      }
    }
  }
};
```

<template v-slot:caption>
/modules/@apostrophecms/global/index.js
</template>
</AposCodeBlock>

### Adding the footer

> Much like we constructed the navigation, we are going to use a fragment populated with data from the global settings for the footer. Create a `views/fragments/footer.html` file with a fragment block and paste the navigation area from any of the template pages between the blocks. In this case, we are going to replace each of the social links and the copyright text. You can choose to make the link for each social media account be required, or wrap each of the list items in an `if` block to make them optional.

Author: Changed and added a note about your alternative.

<AposCodeBlock>

```js
module.exports = {
  fields: {
    add: {
      // ...
      twitterUrl: {
        type: 'url',
        label: 'Twitter URL'
      },
      facebookUrl: {
        type: 'url',
        label: 'Facebook URL'
      },
      githubUrl: {
        type: 'url',
        label: 'Github URL'
      },
      copyright: {
        type: 'string',
        label: 'Copyright text'
      }
    },
    group: {
      // ...
      footer: {
        label: 'Footer URLs and text',
        fields: ['twitterUrl', 'facebookUrl', 'githubUrl', 'copyright' ]
      }
```

<template v-slot:caption>
modules/@apostrophecms/global/index.js
</template>
</AposCodeBlock>

### Adding the header

The headers of each page have an image and headline in common. They also each have a subheading, but the styling of that subheading depends on the type of page that is being displayed. Additionally, the header for the page displaying the individual blog articles also has metadata about the author and publication date. While we could have separate header template fragments, we can also add a conditional block to add the needed markup.

> Create a `views/fragments/header.html` file and add the fragment block tags. Unlike the navigation and footer fragments that got their data from the global settings, the header will get its data from the page. This means that we will have to pass data into our fragment within the block tags - `{% fragment headerArea(data) %}`.

Author: Changed.

:::

The first modification we are going to make is to import our two fragments.

<AposCodeBlock>
```django
{% import 'fragments/navigation.html' as navigation %}
{% import 'fragments/footer.html' as footer %}
```

</AposCodeBlock>

This will bring our fragments into the template and name them `navigation` and `footer`.

To load our font files we are going to take advantage of the `extraHead` section of the `outerLayout` template that our `layout.html` is extending. You can read the [documentation](https://v3.docs.apostrophecms.org/guide/layout-template.html#layout-templates) to learn about other sections of this template that can be extended. Add the `{% block extraHead %}{% endblock %}` tags after the `{% block title%}` section. Within those tags copy the fonts section of the head from any of the template pages. You can also see that there are links for the site favicon and some other meta tags in this section of the templates. You can elect to add those if you desire.

> Finally, delete all of the markup in the `beforeMain` and `afterMain` sections. into the `beforeMain` block render the navigation - `{% render navigation.navigationArea() %}`. In the `afterMain` block render the footer fragment = `{% render footer.footerArea() %}`. The final modification we can make to the template is to add the semantic `<main></main>` tags. Looking at the original template pages, we can see that the main section has a class of `mb-4`. This markup can be added around the `{% block main %}{% endblock %}` tags.

Reviewer: ```suggestion
Finally, delete all of the markup in the `beforeMain` and `afterMain` sections. In the `beforeMain` block, render the navigation - `{% render navigation.navigationArea() %}`. In the `afterMain` block render, the footer fragment - `{% render footer.footerArea() %}`. The final modification we can make to the template is to add the semantic `<main></main>` tags. Looking at the original template pages, we can see that the main section has a class of `mb-4`. This markup can be added around the `{% block main %}{% endblock %}` tags.
```
Formatting.

> This markup can be added around the `{% block main %}{% endblock %}` tags.

Huh. I usually place it within the `main` block. Didn't think to put it outside. Nice.

  {{ title }}
  {% if not title %}
    {{ apos.log('Looks like you forgot to override the title block in a template that does not have access to an Apostrophe page or piece.') }}
  {% endif %}
{% endblock %}

{% block extraHead %}
  <!-- Font Awesome icons (free version)-->
  <script src="https://use.fontawesome.com/releases/v6.1.0/js/all.js" crossorigin="anonymous"></script>
  <!-- Google fonts-->
  <link href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" rel="stylesheet" type="text/css" />
  <link href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" rel="stylesheet" type="text/css" />
{% endblock %}

{% block beforeMain %}
  {% render navigation.navigationArea() %}
{% endblock %}

<main class="mb-4">
{% block main %}
  {#
    Usually, your page templates in the @apostrophecms/pages module will override
    this block. It is safe to assume this is where your page-specific content
    should go.
  #}
{% endblock %}
</main>

{% block afterMain %}
  {% render footer.footerArea() %}
{% endblock %}

```
:::
<template v-slot:caption>
views/layout.html
</template>
</AposCodeBlock>

> Note the use of `navigationArea()` and `footerArea()` in the `render` calls. If you look at the fragment files, you will see that the fragment blocks use these names. This is to allow for a single fragment file to have multiple fragments. We are using these names just to be a little clearer in our calls, but they could just be named `navigation()` and `footer()`.

Reviewer: > We are using these names just to be a little clearer in our calls, but they could just be named `navigation()` and `footer()`.

Probably just me, but I usually name all my fragments consistently so you don't have to remember the name of the render fn from the fragment. Here could just be `area()` so after you do your imports you'd have `navigation.area()` and `footer.area()`. Also, reads more naturally that way. Maybe could just be a note somewhere here since that change might make your fragment examples above a little confusing to a newbie.

Author: Changed the last sentence to reflect how you do this.

## Creating a default page

Now all of our pages will have our navigation and footer areas, but we need to add our header and all of the body content. We could make a separate file for each page of our site, but it makes sense to have a default page that will be used for the non-blog pages. We can use the existing `modules/default-page`.

We need to make two major modifications to the existing `default-page`. First, we need to import and display our header fragment. In this case, we want to *add* the header into the `beforeMain` block. In order to accomplish this we need to add the 

> ::: v-pre

Reviewer: Does this need the `v-pre`? Makes a bit of a weird break on the rendered page.

Author: Yeah, we use VuePress. This allows you to do a lot of templating things with Vue components, but it also means that you can't use double mustaches, even inside of backticks - Grrr!

Reviewer: Booo 👻

`{{ super() }}`
:::
call to include all of the `beforeMain` content from our layout template.

Second, we need to add the template styling and the main content area to the `main` block.

<AposCodeBlock>

::: v-pre

```django
{% extends "layout.html" %}

{% import "fragments/header.html" as header %}

{% block beforeMain %}
  {{ super() }}
  {% render header.headerArea(data.page) %}
{% endblock %}

>   {% block main %}

Reviewer: Extra indentation on this block.

    <div class="container px-4 px-lg-5">
      <div class="row gx-4 gx-lg-5 justify-content-center">
        <div class="col-md-10 col-lg-8 col-xl-7">
          {% area data.page, 'main' %}
        </div>
      </div>
    </div>
  {% endblock %}
```

:::

<template v-slot:caption>
modules/default-page/views/page.html
</template>
</AposCodeBlock>

> Next, we need to modify the schema fields of the default-page `index.js` file to add the data to the header and the main body of the page.

Reviewer: ```suggestion
Next, we need to modify the schema fields of the `default-page/index.js` file to add the data to the header and the main body of the page.
```

`module/file` is called out elsewhere, should that be the same here?

Author: Yup, changed.

        }
      },
      main: {
        type: 'area',
        options: {
          widgets: {
            '@apostrophecms/rich-text': {
              toolbar: [
                'styles',
                '|',
                'bold',
                'italic',
                'strike',
                'link',
                '|',
                'bulletList',
                'orderedList'
              ],
              styles: [
                {
                  tag: 'p',
                  label: 'Paragraph (P)'
                },
                {
                  tag: 'h3',
                  label: 'Heading 3 (H3)'
                },
                {
                  tag: 'h4',
                  label: 'Heading 4 (H4)'
                }
              ]
            },
            '@apostrophecms/image': {},
            '@apostrophecms/video': {}
          }
        }
      }
    },
    group: {
      basics: {
        label: 'Basics',
        fields: [
          'title',
          'heading',
          'subheading',
          'headerImage',
          'main'
        ]
      }
    }
  }
};

```

<template v-slot:caption>
modules/default-page/index.js
</template>
</AposCodeBlock>

> To accommodate the 'Contact Us' page, we could also the widgets from the [form extension](https://apostrophecms.com/extensions/form-builder-3-x) in the `main` area. Since we are using the `default-page` which already exists in our project we don't need to modify either the `app.js` or `modules/@apostrophecms/page/index.js` files.

Reviewer: ```suggestion
To accommodate the 'Contact Us' page, we could also add the widgets from the [form extension](https://apostrophecms.com/extensions/form-builder-3-x) in the `main` area. Since we are using the `default-page` which already exists in our project we don't need to modify either the `app.js` or `modules/@apostrophecms/page/index.js` files.
```

> Since we are using the `default-page` which already exists in our project we don't need to modify either the `app.js` or `modules/@apostrophecms/page/index.js` files.

This sentence tripped me up a couple times. Got what I think you're saying after re-reading a couple times but wasn't completely clear first time around.

Author: Hopefully clarified.

### Modifying the logged-in page display
If we were to take a look at our page right now while logged-in as an editor, we would see a couple of problems. First, the navigation section is styled to be added at the top of the page using a `postion: absolute` CSS rule. The problem with this is that this ends up putting our navigation *over* the ApostropheCMS admin-bar. Not only can we not see the navigation, but this also blocks access to the admin-bar menus. So, we need to add some code onto the page that will move our navigation below the admin-bar in the page flow.

> There are several areas in our project where we could add code to solve this problem. In this case, we will add a small script to our asset module again. While we could add it to `modules/@apostrophecms/asset/ui/src/index.js` along with the template code, this would result in the delivery of extra unnecessary JavaScript to all users. Instead, we will add the code into `modules/@apostrophecms/asset/ui/apos/apps`. This folder is commonly used in projects to add new custom Vue UI components.

Reviewer: ```suggestion
There are several areas in our project where we could add code to solve this problem. In this case, we will add a small script to our asset module again. While we could add it to `modules/@apostrophecms/asset/ui/src/index.js` along with the template code, this would result in the delivery of extra unnecessary JavaScript to all users. Instead, we will add the code into `modules/@apostrophecms/asset/ui/apos/apps`. This folder is commonly used in projects to add new custom Vue UI components and is only served to logged in users.
```

Maybe add a little more detail here so it's clear why `apos/apps` vs `ui/src`.

Author: Changed.

modules/blog/index.js
</template>
</AposCodeBlock>

The `index.js` file for the `blog-page` will also be quite similar. Again, it will have a header image, heading, and subheading.

<AposCodeBlock>

```js
module.exports = {
  extend: '@apostrophecms/piece-page-type',
  options: {
    perPage: 5
  },
  fields: {
    add: {
      headerImage: {
        type: 'area',
        label: 'Header image',
        required: true,
        options: {
          widgets: {
            '@apostrophecms/image': {}
          },
          max: 1
        }
      },
      heading: {
        type: 'string',
        label: 'Heading'
      },
      subheading: {
        type: 'string',
        label: 'Subheading'
      }
    },
    group: {
      basics: {
        label: 'Basics',
        fields: [
          'title',
          'headerImage',
          'heading',
          'subheading'
        ]
      }
    }
  }
};

```

<template v-slot:caption>
modules/blog-page/index.js
</template>
</AposCodeBlock>

In the code above we are adding a `perPage` option of `5`. This will limit the number of blog articles shown on the "Home" page. This can be adjusted to best serve your needs.

### The blog index.html page

> The "Home" page of the template is essentially an `index.html` page that lists all of the blog articles. Just like with the default page, we are going to add our header to a `beforeMain` block. Within the `main` block, we will copy the `Main Content` section from the `index.html` HTML template. To convert it to dynamically show all of the blog articles from our site we will delete all of the code in each of the `Post preview` sections except the first. Then, we will wrap the first `Post preview` section in a `for` loop. Finally, we will modify the `Pager` section to show both newer and older posts.

Reviewer: ```suggestion
The "Home" page of the template is essentially an `index.html` page that lists all of the blog articles. Just like with the default page, we are going to add our header to a `beforeMain` block. Within the `main` block, we will copy the "Main Content" section from the `index.html` HTML template. To convert it to dynamically show all of the blog articles from our site we will delete all of the code in each of the "Post preview" sections except the first. Then, we will wrap the first "Post preview" section in a `for` loop. Finally, we will modify the "Pager" section to show both newer and older posts.
```

Not sure if those should be formatted as code.

Author: I want people to find them in the code, so I added the rest of the comment formatting.

<AposCodeBlock>
::: v-pre

```django
{% extends 'layout.html' %}

{% import 'fragments/header.html' as header %}

{% block beforeMain %}
  {{ super() }}
  {% render header.headerArea(data.page) %}
{% endblock %}

{% block main %}
  <!-- Main Content-->
  <div class="container px-4 px-lg-5">
    <div class="row gx-4 gx-lg-5 justify-content-center">
      <div class="col-md-10 col-lg-8 col-xl-7">
        {% for piece in data.pieces %}
          <!-- Post Preview -->
          <div class="post-preview">
            <a href="{{ piece._url }}">
              <h2 class="post-title">{{ piece.heading }}</h2>
              <h3 class="post-subtitle">{{ piece.subheading }}</h3>
            </a>

>             <p class="post-meta"> Posted by {{ piece.author }} on {{ piece.publicationDate | date('MMMM d, YYYY') }}</p>

Reviewer: Most of the fields in the schema aren't required for the header or the piece type. Should more of them be required with conditionals in the template for the one's that aren't?

Author: Yeah - didn't think that through! I went back and made all the fields required.

          </div>
          <!-- Divider-->
          <hr class="my-4" />
        {% endfor %}
        <!-- Pager-->
        <div class="d-flex mb-4">
          {% if data.currentPage > 1 %}
          <a href="{{ data.slug }}?page={{data.currentPage - 1}}" class="btn btn-primary text-uppercase me-auto">newer posts</a>
          {% endif %}
          {% if data.totalPages > data.currentPage %}
          <a href="{{ data.slug }}?page={{data.currentPage + 1}}" class="btn btn-primary text-uppercase ms-auto">older posts</a>
          {% endif %}
        </div>
      </div>
    </div>
  </div>

{% endblock %}
```
:::

<template v-slot:caption>
modules/blog-page/index.html
</template>
</AposCodeBlock>

Focusing on the `for` loop in the code. We are stepping through all of the articles returned in `data.pieces` and outputing the relevant data. Again, since we specified a `perPage` value of `5` in the options, this will return the five newest blog articles. This can be further configured within the `blog-page` module options, for example with the [`sort` option](https://v3.docs.apostrophecms.org/reference/modules/piece-type.html#sort).

> The `Pager` section is expanded to conditionally show newer and older blog articles, unlike the original template, which only shows older articles. Within the section, we are taking advantage of some additional data that is being delivered to the `index.html` page. Within the `data` payload are `data.currentPage` and `data.totalPages`. The `data.totalPages` is how many individual data sets are present for the particular piece type if divided into groups based on the `perPage` option (the default is 10).

Reviewer: ```suggestion
The "Pager" section is expanded to conditionally show newer and older blog articles, unlike the original template, which only shows older articles. Within the section, we are taking advantage of some additional data that is being delivered to the `index.html` page. Within the `data` payload are `data.currentPage` and `data.totalPages`. The `data.totalPages` is how many individual data sets are present for the particular piece type if divided into groups based on the `perPage` option (the default is `10`).
```

Same as above. Also, value in above paragraph is formatted, maybe should be here too.

Author: Changed.

By default, we are showing the newest blog articles first. Therefore, if the `data.currentPage` is equal to `1` then we shouldn't display the button to load newer articles. If we are on any other page we want the button displayed, with a URL that adds a query to go to the previous page -

> ::: v-pre

Reviewer: ```suggestion
```

Do you need `::: v-pre` here?

Author: Yeah - see comment on other `::: v-pre` section.

> (`{{data.currentPage - 1}}`).

Reviewer: ```suggestion
`{{data.currentPage - 1}}`.
```

Are the parens necessary?

Author: I guess not since it is being bumped onto its own line by the ::: pre. Changed.

:::

> We are only displaying the button to go to older posts if we aren't at the last set of pieces - (`data.totalPages > data.currentPage`). If this is true we display a button that points to the URL with a query that goes to the next set of pieces - 

Reviewer: ```suggestion
We are only displaying the button to go to older posts if we aren't at the last set of pieces - `data.totalPages > data.currentPage`. If this is true we display a button that points to the URL with a query that goes to the next set of pieces - 
```

Are the parens necessary?

Author: Changed.

> ::: v-pre

Reviewer: ```suggestion
```

Same as previous comment. Creates funny breaks in the rendered page.

> (`{{data.currentPage + 1}}`).

Reviewer: ```suggestion
`{{data.currentPage + 1}}`.
```

Are the parens necessary?

Author: Changed.

:::

### The blog show.html page

The `show.html` page will display each of the individual blog articles and will be based on the original HTML template `post.html` page. Like the other pages, we start by rendering the header fragment in the `beforeMain` block. If we look at the `post.html` page we can see that the header looks slightly different from the other pages. It contains metadata not found on the other pages. After setting up the main part of the page we will alter the `header.html` fragment to address this.

> Open the `post.html` template file and copy the `Post Content` section into the `main` block of the `show.html` page. All of the content in `<p>` tags can be deleted because we will replace it with the content added to the `main` area of our blog pieces.

Reviewer: ```suggestion
Open the `post.html` template file and copy the "Post Content" section into the `main` block of the `show.html` page. All of the content in `p` tags can be deleted because we will replace it with the content added to the `main` area of our blog pieces.
```

Same formatting suggestion as above. Not sure you need the `< >` if you also say "tags".

Author: Changed.

<AposCodeBlock>
::: v-pre
```django
{% extends "layout.html" %}

{% import 'fragments/header.html' as header %}

{% block beforeMain %}
    {{ super() }}
    {% render header.headerArea(data.piece) %}
{% endblock %}

{% block main %}
    <!-- Post Content-->
    <article class="mb-4">
        <div class="container px-4 px-lg-5">
            <div class="row gx-4 gx-lg-5 justify-content-center">
                <div class="col-md-10 col-lg-8 col-xl-7">

>                     {% area data.piece,
>                     'main' %}

Reviewer: ```suggestion
                    {% area data.piece, 'main' %}
```

Extra break.

Author: Changed.

<AposCodeBlock>

```django
{% fragment headerArea(data) %}
{% set background = apos.image.first(data.headerImage) %}

{% set type = data.type %}

<!-- Page Header-->
{% if background %}
{% set backgroundUrl = apos.attachment.url(background, { size: 'max' }) %}
<header class="masthead" style="background-image: url('{{ backgroundUrl }}')">
{% endif %}
  <div class="container position-relative px-4 px-lg-5">
    <div class="row gx-4 gx-lg-5 justify-content-center">
      <div class="col-md-10 col-lg-8 col-xl-7">
        <div class={{ "post-heading" if type === 'blog' else "site-heading"}}>
          <h1>{{ data.heading }}</h1>
          {% if type === 'blog' %}
          <h2 class="subheading">{{ data.subheading }}</h2>
          <span class="meta">Posted by <a href="{{ data.authorLink }}">{{ data.author }}</a> on {{ data.publicationDate }} </span>
          {% else %}
          <span class="subheading">{{ data.subheading }}</span>
          {% endif %}
        </div>
      </div>
    </div>
  </div>
</header>
{% endfragment %}
```

<template v-slot:caption>
views/fragments/header.html
</template>
</AposCodeBlock>

> So, what did we change? First, at the top, we created a new variable, `type`, and set it to the value of `data.type`. This is going to give us the type of page/piece that is being loaded. Looking at the code differences in the header section of the `post.html` template compared to the other three pages, the first thing we see is that the container for the header text has a class of `post-heading` instead of `site-heading`. In the code above, we conditionally set the class only if the page being loaded is of the type `blog`. Similarly, we only add the metadata if we are on this type of page.

Reviewer: ```suggestion
So, what did we change? First, at the top, we created a new variable, `type`, and set it to the value of `data.type`. This is going to give us the type of page or piece that is being loaded. Looking at the code differences in the header section of the `post.html` template compared to the other three pages, the first thing we see is that the container for the header text has a class of `post-heading` instead of `site-heading`. In the code above, we conditionally set the class only if the page being loaded is of the type `blog`. Similarly, we only add the metadata if we are on this type of page.
```

I do `thing-a/thing-b` a lot but maybe doesn't make sense here. Not sure if just changing it to `or` is clear but maybe there's a better way to word this.

Author: Changed.

> Now all that is left to do is add our pages to the site and give them some content. The "About Me" and "Contact Me" both use the `default-page` template. The existing "Home" page should be swapped out for a `blog-page` template. All that is left to do is create your blog articles!

Reviewer: "all that is left to do" is repeated twice here. Maybe should reword.

Author: Changed.

## Summary

> Any pre-made HTML template can be converted for use in Apostrophe through just two or three simple steps. First, add the styling to your Apostrophe project. Next, create a template for each of the template pages that substitutes data from schema fields into each area of the page that you want to edit. If needed, add special piece types and piece page types.

Reviewer: ```suggestion
Any pre-made HTML template can be converted for use in Apostrophe through a couple simple steps. First, add the front end assets to your Apostrophe project. Next, create a template for each of the template pages that substitutes data from schema fields into each area of the page that you want to edit. If needed, add special piece types and piece page types.
```

Maybe don't call out a number. There's 6 steps in this tutorial but yea it's pretty easy. We also added more than styling here. 

Should this also be a list instead?

```
Any pre-made HTML template can be converted for use in Apostrophe through a couple simple steps. 

- Add the front end assets to your Apostrophe project. 
- Create a template for each of the template pages that substitutes data from schema fields into each area of the page that you want to edit. 
- Add special piece types and piece page types.
```

Author: Changed - I liked the look of the bullet list. Thanks!
